engineapi/backup_monitor.go

```diff
@@ -109,6 +123,68 @@ func NewBackupMonitor(logger logrus.FieldLogger,
 	return m, nil
 }
 
+// getBackupVolumeRecurringJobs get recurring jobs in the volume labels and recurring jobs of groups in the volume labels
+func (m *BackupMonitor) getBackupVolumeRecurringJobs(ds *datastore.DataStore, volume *longhorn.Volume) (string, error) {
+	allRecurringJobs, err := ds.ListRecurringJobsRO()
+	if err != nil {
+		m.logger.WithError(err).Warn("Failed to list all recurring jobs")
+		return "", err
+	}
+
+	volumeRecurringJobSpecList := make(map[string]longhorn.RecordVolumeRecurringJob)
+	// distinguish volume labels between job and group.
+	volumeSelectedRecurringJobs := datastore.MarshalLabelToVolumeRecurringJob(volume.Labels)
+	for recurringJobName, recurringJob := range allRecurringJobs {
+		for jobName, job := range volumeSelectedRecurringJobs {
+			if job.IsGroup {
+				// add the recurring job from group or update its groups informtaion
+				if volumeRecurringJobSpecList, err = addRecurringJobFromGroup(recurringJobName, jobName, &recurringJob.Spec, volumeRecurringJobSpecList); err != nil {
+					m.logger.WithError(err).WithFields(logrus.Fields{"recurringjob": recurringJobName, "group": jobName}).Warn("Failed to add the recurring job")
+					return "", err
+				}
+			} else if recurringJobName == jobName {
+				// save a recurring job information if it is not a group.
+				if recordJob, exist := volumeRecurringJobSpecList[recurringJobName]; !exist {
+					volumeRecurringJobSpecList[recurringJobName] = longhorn.RecordVolumeRecurringJob{JobSpec: recurringJob.Spec, FromJob: true, FromGroup: []string{}}
+				} else {
+					// recurring job has been saved by groups and update it from job as well.
+					recordJob.FromJob = true
+					volumeRecurringJobSpecList[recurringJobName] = recordJob
+				}
+			}
+		}
+	}
+	volumeRecurringJobInfo, err := json.Marshal(volumeRecurringJobSpecList)
+	if err != nil {
+		m.logger.WithError(err).Warnf("Marshal volumeRecurringJobInfoMap: %v", volumeRecurringJobSpecList)
+		return "", err
+	}
+
+	return string(volumeRecurringJobInfo), nil
+}
+
+// addRecurringJobFromGroup add recurring jobs in the group and add the group name into the recurring job information
+func addRecurringJobFromGroup(jobName, groupName string, jobSpec *longhorn.RecurringJobSpec, volumeRecurringJobInfo map[string]longhorn.RecordVolumeRecurringJob) (map[string]longhorn.RecordVolumeRecurringJob, error) {
```

Comment:
Rename to `addVolumeRecurringJobInfosFromGroup`.

Reply:
Modified and all related codes. Thanks